client/README.md

```diff
@@ -0,0 +1,225 @@
+            if (!isAuthorized) {
+                listener.onFailure(new OpenSearchStatusException("Current user is not authorized to delete resource: " + resourceId, RestStatus.FORBIDDEN));
+                return;
+            }
+
+            // Authorization successful, proceed with deletion
+            ThreadContext threadContext = transportService.getThreadPool().getThreadContext();
+            try (ThreadContext.StoredContext ignored = threadContext.stashContext()) {
+                deleteResource(resourceId, ActionListener.wrap(deleteResponse -> {
+                    if (deleteResponse.getResult() == DocWriteResponse.Result.NOT_FOUND) {
+                        listener.onFailure(new ResourceNotFoundException("Resource " + resourceId + " not found."));
+                    } else {
+                        listener.onResponse(new DeleteResourceResponse("Resource " + resourceId + " deleted successfully."));
+                    }
+                }, exception -> {
+                    log.error("Failed to delete resource: " + resourceId, exception);
+                    listener.onFailure(exception);
+                }));
+            }
+        }, exception -> {
+            log.error("Failed to verify resource access: " + resourceId, exception);
+            listener.onFailure(exception);
+        })
+    );
+}
+```
+
+---
+
+## **Available Java APIs**
+
+The **`ResourceSharingClient`** provides **four Java APIs** for **resource access control**, enabling plugins to **verify, share, revoke, and list** shareableResources.
+
+**Package Location:**
+[`org.opensearch.security.client.resources.ResourceSharingClient`](../client/src/main/java/org/opensearch/security/client/resources/ResourceSharingClient.java)
+
+---
+
+### **API Usage Examples**
+Below are examples demonstrating how to use each API effectively.
+
+---
+
+### **1. `verifyResourceAccess`**
+**Checks if the current user has access to a resource**.
+
+#### **Method Signature:**
+```java
+void verifyResourceAccess(String resourceId, String resourceIndex, ActionListener<Boolean> listener);
+```
+
+#### **Example Usage:**
+```java
+resourceSharingClient.verifyResourceAccess(
+    "resource-123",
+    "resource_index",
+    ActionListener.wrap(isAuthorized -> {
+        if (isAuthorized) {
+            System.out.println("User has access to the resource.");
+        } else {
+            System.out.println("Access denied.");
+        }
+    }, e -> {
+        System.err.println("Failed to verify access: " + e.getMessage());
+    })
+);
+```
+> **Use Case:** Before performing operations like **deletion or modifications**, ensure the user has the right permissions.
+
+---
+
+### **2. `shareResource`**
+**Grants access to a resource** for specific users, roles, or backend roles.
+
+#### **Method Signature:**
+```java
+void shareResource(String resourceId, String resourceIndex, Map<String, Object> shareWith, ActionListener<ResourceSharing> listener);
```

Comment:
Comment about this and other similar methods - since we know that the keys would be "roles", "users", and "backend_roles", can we somehow enforce that they provide a ShareWith datastructure, instead of accepting an arbitrary map here?

Reply:
This will eventually be replaced with ShareWith class. I can replace it for now to be of type: `SharedWithActionGroup.ActionGroupRecipients recipients`

Reply:
I've made method arguments more strictly typed.